### PR TITLE
Cleanup DateTimeFormatInfo tests

### LINKED ISSUE #2336: Rename tests in System.Globalization
About 40-50% of the tests in System.Globalization do not have self descriptive names. They are currently named PosTest, NegTest etc.

### PR DESCRIPTION
**Pull request by hughbe**
The Ubuntu build might fail, as I don't have an Ubuntu VM on hand.

Other related System.Globalization.Tests PRs may cause merge conflicts, as the csproj file changes often

### CODE REVIEW

**Comment by ellismg** [comment]
@dotnet-bot test Ubuntu15.10 please

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoCalendarWeekRule.cs [hunk 1]
```diff
@@ -11,64 +12,35 @@ namespace System.Globalization.Tests
...
-        {
-            VerificationHelper(DateTimeFormatInfo.InvariantInfo,
-                    CalendarWeekRule.FirstDay,
-                    false);
-        }
-
-        // TestSetter: Call CalendarWeekRule setter method should return correct value
-        [Fact]
-        public void TestSetter()
+        public static IEnumerable<object[]> CalendarWeekRule_TestData()
```
Comment (ellismg) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoCalendarWeekRule.cs]: You could consider just moving to [InlineData] and taking a String as the first argument and doing `new CultureInfo(<arg>).DateTimeFormat`.  In this case, the string for invariant is `""` (the empty string).  Might reduce some of the ceremony here.

**Comment by ellismg** [comment]
Failure looks unrelated (core dump in System.Net.Security.Tests).  I don't think we have a way to extract the cores yet either :-(.

> 21:41:18 ./run-test.sh: line 131: 27039 Aborted                 (core dumped) ./corerun xunit.console.netcore.exe $testDllName -xml testResults.xml -notrait category=failing -notrait category=OuterLoop -notrait category=$xunitOSCategory -notrait Benchmark=true
21:41:18 error: One or more tests failed while running tests from 'System.Net.Security.Tests'.  Exit code 134.

**Comment by ellismg** [comment]
@dotnet-bot test Ubuntu15.10 please

**Comment by ellismg** [comment]
Thanks for doing this @hughbe!  I made one small comment about preferring InlineData in a few places, as I think it makes the code easier to read, but otherwise, LGTM.

**Comment by hughbe** [comment]
@ellismg I don't have a strong opinion on this, but IMO it is clearer to someone reading the code to use InvariantInfo than "". I can change this though if you'd like

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs [hunk 2]
```diff
@@ -10,54 +10,26 @@ namespace System.Globalization.Tests
...
-        private void VerificationHelper(DateTimeFormatInfo info, string expected, bool setter)
-        {
-            if (setter)
-            {
-                info.AMDesignator = expected;
-            }
-
-            string actual = info.AMDesignator;
-            Assert.Equal(expected, actual);
+            Assert.Throws<ArgumentNullException>(() => new DateTimeFormatInfo().AMDesignator = null); // Value is null
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs]: A couple of things here:
* I, personally, don't like trailing comments.  They get eaten by some refactoring tools, and they're usually more helpful explaining what you're about to do, rather than what you just did.
* Consider the potential virtue of verifying that after an invalid set that AMDesignator still has the previous value.
    * For maximum effect it should probably have a custom value.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedDayNames.cs [hunk 3]
```diff
@@ -10,70 +10,30 @@ namespace System.Globalization.Tests
...
-        public void TestNullProperty()
+        public void AbbreviatedDayNames_Set_Invalid()
         {
-            Assert.Throws<ArgumentNullException>(() =>
-            {
-                new DateTimeFormatInfo().AbbreviatedDayNames = null;
-            });
+            Assert.Throws<ArgumentNullException>(() => new DateTimeFormatInfo().AbbreviatedDayNames = null); // Value is null
+            Assert.Throws<ArgumentNullException>(() => new DateTimeFormatInfo().AbbreviatedDayNames = new string[] { "1", "2", "3", null, "5", "6", "7" }); // Value has null
+            Assert.Throws<ArgumentException>(() => new DateTimeFormatInfo().AbbreviatedDayNames = new string[] { "sun" }); // Value.Length is not 13
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedDayNames.cs]: I assume the problem is that value.Length is not 7; as opposed to the stated 13?

Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedDayNames.cs]: Whoops, I'll change this

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs]: I like to use trailing comments only for exception throwing, as it formalizes what I'm doing. I can remove these comments if you want, but I've never been told not to do that before for changes like this in previous PRs

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs [hunk 4]
```diff
@@ -10,112 +10,30 @@ namespace System.Globalization.Tests
...
-                    "7",
-                    "8",
-                    "9",
-                    "10",
-                    "11",
-                    "12",
-                    ""
-                    },
-                    true);
+            string[] newAbbreviatedMonthGenitiveNames = new string[] { "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "" };
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs]: None of the tests set a non-empty value in the last position.  Is empty always required? If so, a negative test would be good.  If not, a positive test would be good.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoCalendarWeekRule.cs (hunk 1 above)
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoCalendarWeekRule.cs]: @ellismg I'm not sure that would cleanly work here since the br-FR case doesn't follow the template for the second element.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs]: FYI: I was told off for adding tests in a precious refactoring PR. I can add some tests in the future but maybe in a future PR? 
If you'd like I can make an issue tracking tests that we should add if you provide me with a list :)

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoClone.cs [hunk 5]
```diff
@@ -10,82 +11,46 @@ namespace System.Globalization.Tests
...
-        public void PosTest3()
-        {
-            DateTimeFormatInfo expected = CultureInfo.InvariantCulture.DateTimeFormat;
-            VerificationHelper(expected, expected.Clone());
-        }
-
-        private void VerificationHelper(DateTimeFormatInfo expected, Object obj)
+        [Theory]
+        [MemberData(nameof(Clone_TestData))]
+        public void Clone(DateTimeFormatInfo format)
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoClone.cs]: `format` doesn't appear to be used?

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoClone.cs [hunk 6]
```diff
@@ -10,82 +11,46 @@ namespace System.Globalization.Tests
...
         {
+            DateTimeFormatInfo expected = new DateTimeFormatInfo();
+            object obj = expected.Clone();
             Assert.True(obj is DateTimeFormatInfo);
+
             DateTimeFormatInfo actual = obj as DateTimeFormatInfo;
             Assert.False(actual.IsReadOnly);
-            IsEquals(actual.AbbreviatedDayNames, expected.AbbreviatedDayNames);
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoClone.cs]: `Assert.NotSame(expected, actual);`, perhaps?

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs]: > I was told off for adding tests in a precious refactoring PR

If you want to add additional test cases, that's fine, but please do it in a separate commit.  When reviewing a refactoring, we need to be able to correlate the old with the new, and that because much more difficult when meaningful test changes are mixed in.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs]: We don't really have a rule on it, so it comes down to personal style.  As with all personal style discrepancies the options are 1) present a logical reason for your style and leave it at that, 2) flame war, 3) bite your tongue.

I tried for (1) :smile:.

Oh, and another reason I don't like them:
* Makes the lines longer
    * Which, combined with github's scrolling rules on large files, makes reviewing things annoying at times.

:smile:

So, it's really just food for thought; not a must-fix.

Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAMDesignator.cs]: ahh good reason ;)

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoAbbreviatedMonthGenitiveNames.cs]: I understand the reasons perfectly! :D

**Comment by hughbe** [comment]
Pushed a PR feedback commit

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoDayNames.cs [hunk 7]
```diff
@@ -10,92 +10,30 @@ namespace System.Globalization.Tests
...
-
-        private void VerificationHelper(DateTimeFormatInfo info, string[] expected, bool setter)
-        {
-            if (setter)
-            {
-                info.DayNames = expected;
-            }
+            Assert.Throws<ArgumentNullException>(() => new DateTimeFormatInfo().DayNames = null); // Value is null
+            Assert.Throws<ArgumentNullException>(() => new DateTimeFormatInfo().DayNames = new string[] { "1", "2", "3", null, "5", "6", "7" }); // Value has null
+            Assert.Throws<ArgumentException>(() => new DateTimeFormatInfo().DayNames = new string[] { "sun" }); // Value.Length is not 13
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoDayNames.cs]: Another 13 v 7 comment discrepancy.  Maybe we should just reshape the week to fit... so long as we keep a 5 day work week... nah, we'd have to name the new days, and that'll be stuck in committee forever.

Anyways, not going to point out any more of this discrepancy, but with this one so far from the first it seemed like a good "there might be more of these" moment.

Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoDayNames.cs]: Ahahha what a wonderful idea!

Yeah I found this and a couple of similar thinkos and fixed 'em.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetAbbreviatedEraName.cs [hunk 8]
```diff
@@ -11,59 +12,28 @@ namespace System.Globalization.Tests
...
-        }
-
-        private void VerificationHelper(DateTimeFormatInfo info, int era, string expected)
+        public void GetAbbreviatedEraName_Invalid()
         {
-            string actual = info.GetAbbreviatedEraName(era);
-            Assert.Equal(expected, actual);
+            var format = new CultureInfo("en-US").DateTimeFormat;
+            Assert.Throws<ArgumentOutOfRangeException>(() => format.GetAbbreviatedEraName(-1)); // Era < 0
+            Assert.Throws<ArgumentOutOfRangeException>(() => format.GetAbbreviatedEraName(2)); // Era > 1
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetAbbreviatedEraName.cs]: While the magic -1 conveys fairly obvious implicit semantic information; the 2 is much harder.  In an ideal world I'd expect this to be one test for confirming that en-US has two eras, and then this to just request GetEraCount() (not that I know that there's a good way to get the era count)

Comment (hughbe) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetAbbreviatedEraName.cs]: I can extract this 2 as a local, and call it `enUSEraCoun`t, then pass `enUSEraCount + 1` instead of 2

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetFormat.cs [hunk 9]
```diff
@@ -10,26 +10,19 @@ namespace System.Globalization.Tests
...
         {
             DateTimeFormatInfo expected = new DateTimeFormatInfo();
-            object obj = expected.GetFormat(typeof(DateTimeFormatInfo));
-
-            Assert.True(obj is DateTimeFormatInfo);
-
-            DateTimeFormatInfo actual = obj as DateTimeFormatInfo;
-            Assert.Equal(expected, actual);
+            DateTimeFormatInfo format = (DateTimeFormatInfo)expected.GetFormat(typeof(DateTimeFormatInfo));
+            Assert.Equal(expected, format);
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetFormat.cs]: Are there Same-ness guarantees that need to be verified here?

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetMonthName.cs [hunk 10]
```diff
@@ -2,145 +2,44 @@
...
-                "novembre",
-                "d\u00E9cembre",
-                "",
-            };
-
-            VerificationHelper(info, expected);
-        }
+            yield return new object[] { DateTimeFormatInfo.InvariantInfo, new string[] { "", "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December", "" } };
+            yield return new object[] { new DateTimeFormatInfo(), new string[] { "", "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December", "" } };
+            yield return new object[] { new CultureInfo("en-US").DateTimeFormat, new string[] { "", "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December", "" } };
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoGetMonthName.cs]: All three of these use the same data.  Can we create the array just once, and pass it with all three?  This applies in general.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoMonthGenitiveNames.cs [hunk 11]
```diff
@@ -11,134 +12,52 @@ namespace System.Globalization.Tests
...
+                "\u043C\u0430\u044F",
+                "\u0438\u044E\u043D\u044F",
+                "\u0438\u044E\u043B\u044F",
+                "\u0430\u0432\u0433\u0443\u0441\u0442\u0430",
+                "\u0441\u0435\u043D\u0442\u044F\u0431\u0440\u044F",
+                "\u043E\u043A\u0442\u044F\u0431\u0440\u044F",
+                "\u043D\u043E\u044F\u0431\u0440\u044F",
+                "\u0434\u0435\u043A\u0430\u0431\u0440\u044F",
+                ""
+            } };
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoMonthGenitiveNames.cs]: Style nit-pref.   I don't think we have a declarative preferred style here, but I like chopping before the first value, and that keeps indentation levels scoped:

```C#
yield return new object[]
{
    new CultureInfo("ru-RU").DateTimeFormat,
    new string[]
    {
        "\u044F\u043D\u0432\u0430\u0440\u044F",
        ...
    },
};

That also eliminates things like the double-close-curly.

For me it just helps my eyes focus on where the different pieces come from/go.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoReadOnly.cs [hunk 12]
```diff
@@ -10,34 +11,26 @@ namespace System.Globalization.Tests
...
+            yield return new object[] { new CultureInfo("en-US").DateTimeFormat, false };
+            yield return new object[] { DateTimeFormatInfo.ReadOnly(new CultureInfo("en-US").DateTimeFormat), true };
         }
 
-        // PosTest2: Call ReadOnly on a read only DateTimeFormatInfo instance
-        [Fact]
-        public void TestReadOnly()
+        [Theory]
+        [MemberData(nameof(ReadOnly_TestData))]
+        public void ReadOnly(DateTimeFormatInfo format, bool expected)
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoReadOnly.cs]: The calls to the method `ReadOnly` have moved inside the MemberData bundle; but before the refactoring the test seemed more about the `ReadOnly` method, and after the `IsReadOnly` property.

I'd expect this test method to be calling ReadOnly.  Admittedly, some of the input may have already called ReadOnly.

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs [hunk 13]
```diff
@@ -10,54 +11,35 @@ namespace System.Globalization.Tests
 {
     public class DateTimeFormatInfoShortDatePattern
     {
-        private readonly RandomDataGenerator _generator = new RandomDataGenerator();
+        private static readonly RandomDataGenerator s_RandomDataGenerator = new RandomDataGenerator();
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs]: Naming: `s_randomDataGenerator` (pervasive)

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoYearMonthPattern.cs [hunk 14]
```diff
@@ -10,55 +11,43 @@ namespace System.Globalization.Tests
...
-            VerificationHelper(DateTimeFormatInfo.InvariantInfo, "yyyy MMMM", false);
-            VerificationHelper(new CultureInfo("fr-fr").DateTimeFormat, "MMMM yyyy", false);
+            yield return new object[] { DateTimeFormatInfo.InvariantInfo, "yyyy MMMM" };
+            yield return new object[] { new CultureInfo("fr-FR").DateTimeFormat, "MMMM yyyy" };
         }
-
-        // PosTest2: Call YearMonthPattern setter method should return correct value
-        [Fact]
-        public void TestSetter()
+
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoYearMonthPattern.cs]: Nit: Extraneous end-of-line whitespace.

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs [hunk 15]
```diff
@@ -11,7 +11,7 @@ namespace System.Globalization.Tests
     internal static class NumberFormatInfoData
     {
         private static readonly bool s_isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
-        private static readonly int s_WindowsVersion = DateTimeFormatInfoData.GetWindowsVersion();
+        private static readonly int s_WindowsVersion = PlatformDetection.WindowsVersion;
         private static readonly bool s_isOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
         private static readonly Version s_OSXKernelVersion = GetOSXKernelVersion();
```
Comment (bartonjs) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs]: Elsewhere you switched things like these to use PlatformDetection; why are these still here?

Comment (hughbe) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs]: I didn't want to change NumberFormatInfo too much in this PR, but beware, I've got a PR coming that's gonna cleanup a whole bunch of stuff in NFI tests

**Comment by hughbe** [comment]
@bartonjs pushed a PR feedback commit responding to all your comments

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoSortableDateTimePattern.cs [hunk 16]
```diff
@@ -10,19 +11,18 @@ namespace System.Globalization.Tests
...
     {
-        // PosTest1: Call SortableDateTimePattern getter method should return correct value for InvariantInfo
-        [Fact]
-        public void TestGetter()
-        {
-            VerificationHelper(DateTimeFormatInfo.InvariantInfo, "yyyy'-'MM'-'dd'T'HH':'mm':'ss");
-            VerificationHelper(new CultureInfo("en-us").DateTimeFormat, "yyyy'-'MM'-'dd'T'HH':'mm':'ss");
-            VerificationHelper(new CultureInfo("ja-JP").DateTimeFormat, "yyyy'-'MM'-'dd'T'HH':'mm':'ss");
+        public static IEnumerable<object[]> SortableDateTimePattern_TestData()
+        {
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoSortableDateTimePattern.cs]: The latest commit brought some more extraneous trailing whitespace in.  `git diff --check` shows unstaged trailing whitespace, `git diff --check --cached` shows staged; and it looks like `git diff HEAD~2 --check` will show it for your last two commits

File: src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoMonthGenitiveNames.cs [hunk 17]
```diff
@@ -11,59 +12,13 @@ namespace System.Globalization.Tests
...
-        [Fact]
-        public void TestRuCulture()
-        {
-            DateTimeFormatInfo info = new CultureInfo("ru-RU").DateTimeFormat;
-
-            string[] expected = new string[] {
+            yield return new object[] { DateTimeFormatInfo.InvariantInfo, new string[] { "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December", "" } };
+            yield return new object[] 
+            {
+                new CultureInfo("ru-RU").DateTimeFormat, new string[]
```
Comment (bartonjs) [inline src/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoMonthGenitiveNames.cs]: I'd also chop it between the elements here, so it's more visible that the CultureInfo and the string array are different; but at least now the indentation level pops :)

**Comment by bartonjs** [comment]
LGTM assuming squash.  @stephentoub or @ellismg, anything to add?

(Also, as Matt already said: Thanks, @hughbe)

**Comment by hughbe** [comment]
Anytime boss. Can you clarify what you mean by assuming squash?

**Comment by hughbe** [comment]
@bartonjs fixed those final issues :)

**Comment by bartonjs** [comment]
> Can you clarify what you mean by assuming squash?

Please squash the PR feedback commit into the base commit

> @bartonjs fixed those final issues :)

Danke :)

**Comment by hughbe** [comment]
I've squashed the commits, thanks
